File: CPP/router.cpp
#include "router.h"

// =================================================================================================

#ifdef _DEBUG
#   define LOG(msg, ...) fprintf(stderr, msg, ##__VA_ARGS__);
#else
#   define LOG(msg, ...)
#endif

// ================================================================================
// Dial heap for DACS path finding in graphs
// costIndex contains the total path cost to each node stored in it
// Add nodes into costIndex at the position of their predecessor in costIndex plus their edge cost
// Each costIndex entry has a list of nodes with the same path cost
// Next node to expand the path from is always the next node in costIndex seen from the current costIndex position
// Path cost can and will wrap around costIndex; that's why costIndex needs to be larger than the highest
// edge cost.
// See the internets for a full explanation of DACS (Dijkstra Address Calculation Sort)
// nodeLists contains the root indices for lists of nodes with the same cost. These indices point into nodeListLinks.
// nodeLists is indexed with path costs, nodeListLinks is indexed with node ids. For each node id stored in it, it 
// holds the index of the next node id in the list at the index of the next node's predecessor.
//
//   0   1   2         // of nodes - 1
// +---+---+---+-   -+---+
// |   | 5 |   | ... |   |
// +---+---+---+-   -+---+
//       |
//       +----------------+
//           +--+ +-----+ |
//           |  | |     | |
//           v  | v     | v
// +---+---+---+---+---+---+---+---+---+
// |   |   |-1 | 2 |   | 3 |   |   |   | ...
// +---+---+---+---+---+---+---+---+---+
//   0   1   2   3   4   5   6   7   8
//
// This means that 1 there is a list of nodes with path cost 1. nodeList [1] contains 5, that's the 1st node in the list.
// nodeListLinks [5] contains 3, which is the 2nd node in the list. nodeListLinks [3] contains 2, and nodeListLinks [2]
// contains -1, so 2 is the last node in the list of nodes with cost 1.
// The resulting list is 5,3,2

void CDialHeap::Create(int maxNodes) {
    m_maxNodes = maxNodes;
    m_nodeLists.Create(65536);
    m_nodeLists.Fill(-1);
    m_nodeListLinks.Create(m_maxNodes);
    m_nodeListLinks.Fill(-1);
    m_predecessors.Create(m_maxNodes);
    m_predecessors.Fill(-1);
    m_pathCost.Create(m_maxNodes);
    m_pathCost.Fill(65535);
    m_finalCost.Create(m_maxNodes);
    m_finalCost.Fill(65535);
    m_edges.Create(m_maxNodes);
    m_edges.Clear(0);
    m_dirtyIndex.Create(65536);
    m_dirtyCost.Create(65536);
    m_dirtyFinalCost.Create(65536);
}


void CDialHeap::Destroy(void) {
    m_maxNodes = 0;
    m_nodeLists.Destroy();
    m_nodeListLinks.Destroy();
    m_predecessors.Destroy();
    m_pathCost.Destroy();
    m_finalCost.Destroy();
    m_edges.Destroy();
    m_dirtyIndex.Destroy();
    m_dirtyCost.Destroy();
    m_dirtyFinalCost.Destroy();
}


// reset all used list data
void CDialHeap::Reset(void) {
    while (m_dirtyIndex.ToS())
        m_nodeLists[m_dirtyIndex.Pop()] = -1;
    while (m_dirtyCost.ToS())
        m_pathCost[m_dirtyCost.Pop()] = 65535;
    while (m_dirtyFinalCost.ToS())
        m_finalCost[m_dirtyFinalCost.Pop()] = 65535;
    m_costIndex = 0;
}


// start path finding for node node
void CDialHeap::Setup(int nodeId) {
    Reset();
    Push(nodeId, -1, -1, 0);
}


// put the current node node with path cost newCost in the heap or update its cost
// if it is already in the heap and has higher path cost
bool CDialHeap::Push(int nodeId, int predNodeId, int edgeId, uint16_t newCost) {
    uint16_t oldCost = m_pathCost[nodeId];
    if (newCost >= oldCost) 
        return false;    // new path is longer than the currently stored one

    uint16_t costIndex = newCost;
    if (oldCost == 65535) 
        m_dirtyCost.Push (nodeId);
    else {
        // node already in heap with higher pathCost, so unlink node from node list at current path cost position
        // find the node in the node list attached to the nodeList at the current path cost and let its successor take its place in the list
        uint16_t listRoot = oldCost;
        int16_t currNodeId = m_nodeLists[listRoot];
        int16_t nextNodeId = -1;
        while (currNodeId >= 0) {
            if (currNodeId == nodeId) {
                if (nextNodeId < 0)
                    m_nodeLists[listRoot] = m_nodeListLinks[currNodeId];
                else
                    m_nodeListLinks[nextNodeId] = m_nodeListLinks[currNodeId];
                break;
            }
            nextNodeId = currNodeId;
            currNodeId = m_nodeListLinks[currNodeId];
        }
    }

    if (0 > m_nodeLists[costIndex]) 
        m_dirtyIndex.Push(costIndex);
    m_pathCost[nodeId] = newCost;
    m_predecessors[nodeId] = predNodeId;
    m_nodeListLinks[nodeId] = m_nodeLists[costIndex];
    m_nodeLists[costIndex] = nodeId;
    m_edges[nodeId] = edgeId;
    return true;
}


// find node with lowest path cost from current path finding state by searching through the
// node cost offset table from the current node's position there to the next position in the table
// holding a node
int CDialHeap::Scan(int nStart) {
    int l = int(m_nodeLists.Length());
    int i = nStart;
    int j = l;
    while (j-- > 0) {
        if (m_nodeLists[i] >= 0)
            return i;
        ++i %= l;
    }
    return -1;
}


// remove node from path finding tree
auto CDialHeap::Pop(void) {
    struct retVals {
        int nodeId, cost;
    };

    int i = Scan(m_costIndex);
    if (i < 0)
        return retVals{ -1, -1 };
    m_costIndex = i;
    uint16_t nodeId = m_nodeLists[m_costIndex];
    m_nodeLists[m_costIndex] = m_nodeListLinks[nodeId];
    uint16_t cost = m_pathCost[nodeId];
    m_finalCost[nodeId] = cost;
    m_dirtyFinalCost.Push(nodeId);
    return retVals{ nodeId, cost };
}


// calculate length in segments from a node to the start node
int CDialHeap::RouteNodeCount(int nodeId) {
    // i = 65535
    // while (i > 0) {
    //     nodeId = m_predecessors [nodeId]
    //     if (h < 0) {
    //         break
    // return 65536 - i

    int i = 0;
    for (;;) {
        i++;
        if (0 > (nodeId = m_predecessors[nodeId]))
            break;
    }
    return i;
}


CList<CRouteNode>& CDialHeap::BuildRoute(int nodeId) {
    m_route.Destroy();
    for (;;) {
        m_route.Insert(0, CRouteNode(nodeId, m_edges[nodeId]));
        nodeId = m_predecessors[nodeId];
        if (nodeId < 0)
            break;
    }
    return m_route;
}


// ================================================================================
// unidirectional Dijkstra address calculation sort

CList<CRouteNode>& CRouter::BuildPath(int segmentId) {
    m_route.Destroy();
    if (m_pathCost[segmentId] != 65535)
        BuildRoute(segmentId);
    return m_route;
}


// find a path from segment startSegId to segment destSegId in a game map
// if destSegId is -1, compute minimal path cost to each other segment in the map
// that is reachable from the start segment
int CRouter::FindPath(int startSegId, int destSegId, CSegmentMap& segmentMap) {
    Setup(startSegId);
    int expanded = 1;

    LOG ("\nfinding path from %d to %d\n", startSegId, destSegId)
    for (;;) {
        auto [segId, dist] = Pop();
        LOG ("%5d: %5d\n", segId, dist)
        if (segId < 0)
            return (destSegId < 0) ? expanded : -1;
        if (segId == destSegId)
            return int (BuildPath(segId).Length());

        CMapSegment& segment = segmentMap[segId];
        for (auto [i, edgeId] : segment.m_pathEdgeIds) {
            CSegmentPathEdge e = segmentMap.m_pathEdgeTable [edgeId];
            if (Push(e.m_segmentId, segId, edgeId, dist + e.m_distance))   
                expanded++;
        }
    }
}

CRouter router;

// ================================================================================

File: Python/router.py
import numpy as np

from collections import deque
from queue import LifoQueue
from mapsegments import *
from map import *

# ================================================================================

class CPathNode:
    def __init__ (self, node = -1, edge = -1):
        self.node = node
        self.edge = edge

# ================================================================================
# Dial heap for DACS path finding in graphs
# costIndex contains the total path cost to each node stored in it
# Add nodes into costIndex at the position of their predecessor in costIndex plus their edge cost
# Each costIndex entry has a list of nodes with the same path cost
# Next node to expand the path from is always the next node in costIndex seen from the current costIndex position
# Path cost can and will wrap around costIndex; that's why costIndex needs to be larger than the highest
# edge cost.
# See the internets for a full explanation of DACS (Dijkstra Address Calculation Sort)
# nodeLists contains the root indices for lists of nodes with the same cost. These indices point into nodeListLinks.
# nodeLists is indexed with path costs, nodeListLinks is indexed with node ids. For each node id stored in it, it 
# holds the index of the next node id in the list at the index of the next node's predecessor.
#
#   0   1   2         # of nodes - 1
# +---+---+---+-   -+---+
# |   | 5 |   | ... |   |
# +---+---+---+-   -+---+
#       |
#       +----------------+
#           +--+ +-----+ |
#           |  | |     | |
#           v  | v     | v
# +---+---+---+---+---+---+---+---+---+
# |   |   |-1 | 2 |   | 3 |   |   |   | ...
# +---+---+---+---+---+---+---+---+---+
#   0   1   2   3   4   5   6   7   8
#
# This means that 1 there is a list of nodes with path cost 1. nodeList [1] contains 5, that's the 1st node in the list.
# nodeListLinks [5] contains 3, which is the 2nd node in the list. nodeListLinks [3] contains 2, and nodeListLinks [2]
# contains -1, so 2 is the last node in the list of nodes with cost 1.
# The resulting list is 5,3,2

class CDialHeap:
    def __init__ (self):
        self.nodeLists = None
        self.dirtyIndex = None      # list of used nodeLists table entries
        self.nodeListLinks = None       # list of nodes with same position in costIndex
        self.predecessors = None
        self.edges = None
        self.pathCost = None
        self.finalCost = None
        self.dirtyCost = None       # list of used pathCost table entries
        self.dirtyFinalCost = None
        self.route = []
        self.maxNodes = int (0)
        self.costIndex = int (0)
        self.noCost = 65535
        self.maxCost = self.noCost - 1


    def Create (self, maxNodes):
        self.maxNodes = int (maxNodes)
        self.nodeLists = np.full (65536, int (-1), np.int16)
        self.pathCost = np.full (self.maxNodes, int (65535), np.uint16)
        self.finalCost = np.full (self.maxNodes, int (65535), np.uint16)
        self.nodeListLinks = np.full (self.maxNodes, int (-1), np.int16)
        self.predecessors = np.full (self.maxNodes, int (-1), np.int16)
        self.edges = np.zeros (self.maxNodes, np.int16)
        self.dirtyIndex = LifoQueue (maxsize = 65536)
        self.dirtyCost = LifoQueue (maxsize = 65536)
        self.dirtyFinalCost = LifoQueue (maxsize = maxNodes)


    def Destroy (self):
        self.maxNodes = 0


    # reset all used list data
    def Reset (self):
        while not self.dirtyIndex.empty():
            self.nodeLists [self.dirtyIndex.get()] = -1
        while not self.dirtyCost.empty():
            self.pathCost [self.dirtyCost.get()] = 65535
        while not self.dirtyFinalCost.empty():
            self.finalCost [self.dirtyFinalCost.get()] = 65535
        self.costIndex = 0


    # start path finding for node node
    def Setup (self, node):
        self.Reset ()
        self.Push (node, -1, -1, 0)


    # put the current node node with path cost newCost in the heap or update its cost
    # if it is already in the heap and has higher path cost
    def Push (self, node, predNode, edge, newCost):
        oldCost = self.pathCost [node]
        if (newCost >= oldCost):
            return False    # new path is longer than the currently stored one

        costIndex = np.uint16 (newCost)
        if (oldCost == 65535): 
            self.dirtyCost.put (node)
        else:
            # node already in heap with higher pathCost, so unlink
            # unlink node from node list at current path cost position
            # find the node in the node list attached to the nodeList at the current path cost
            # and let its successor take its place in the list
            listRoot = np.uint16 (oldCost)
            currNode = self.nodeLists [listRoot]
            nextNode = -1
            while (currNode >= 0):
                if (currNode == node):
                    if (nextNode < 0):
                        self.nodeLists [listRoot] = self.nodeListLinks [currNode]
                    else:
                        self.nodeListLinks [nextNode] = self.nodeListLinks [currNode]
                    break
                nextNode = currNode
                currNode = self.nodeListLinks [nextNode]

        if (0 > self.nodeLists [costIndex]):
            self.dirtyIndex.put (costIndex)
        self.pathCost [node] = newCost
        self.predecessors [node] = predNode
        self.nodeListLinks [node] = self.nodeLists [costIndex]
        self.nodeLists [costIndex] = node
        self.edges [node] = edge
        return True


    # find node with lowest path cost from current path finding state by searching through the
    # node cost offset table from the current node's position there to the next position in the table
    # holding a node
    def Scan (self, nStart):
        l = len (self.nodeLists)
        i = nStart
        j = l
        while (j > 0):
            if self.nodeLists [i] >= 0:
                return i
            j -= 1
            i += 1
            i %= l
        return -1


    # remove node from path finding tree
    def Pop (self):
        i = self.Scan (self.costIndex)
        if (i < 0):
            return -1, -1
        self.costIndex = i
        node = self.nodeLists [self.costIndex]
        self.nodeLists [self.costIndex] = self.nodeListLinks [node]
        cost = self.pathCost [node]
        #self.pathCost [node] = -1
        self.finalCost [node] = cost
        self.dirtyFinalCost.put (node)
        return node, cost


    # calculate length in segments from a node to the start node
    def RouteNodeCount (self, node):
        h = node

        # i = 65535
        # while (i > 0):
        #     h = self.predecessors [h]
        #     if (h < 0):
        #         break
        # return 65536 - i

        i = 0
        while True:
            i += 1
            h = self.predecessors [h]
            if (h < 0):
                break
        return i


    def BuildRoute (self, node):
        self.route = []
        while True:
            self.route.insert (0, CPathNode (node, self.edges [node]))
            node = self.predecessors [node]
            if (node < 0):
                break
        return self.route
            

    def FinalCost (self, node):
        return self.finalCost [node]


    def Pushed (self, node):
        return self.pathCost [node] < 65535


    def Popped (self, node):
        return not self.Pushed (node) and (self.pathCost [node] < 0)


    def RouteNode (self, i = 0):
        return self.route [i]

# ================================================================================
# unidirectional Dijkstra address calculation sort

class CRouter (CDialHeap):
    def __init__ (self):
        super ().__init__ ()


    def SetSize (self, size):
        self.maxNodes = size


    def BuildPath (self, segmentId):
        self.route = []
        if (self.pathCost [segmentId] != 65535):
            self.BuildRoute (segmentId)
        return self.route


    # find a path from segment startSegId to segment destSegId in a game map
    # if destSegId is -1, compute minimal path cost to each other segment in the map
    # that is reachable from the start segment
    def FindPath (self, startSegId, destSegId, segments):
        dist = 0
        self.Setup (startSegId)
        expanded = 1

        # print ("\nfinding path from {0} to {1}".format (startSegId, destSegId))
        while True:
            segId, dist = self.Pop ()
            if (segId < 0):
                if (destSegId < 0):
                    # print ("expanded {0} nodes".format (expanded))
                    return expanded
                else:
                    return -1
            if (segId == destSegId):
                return self.BuildPath (segId)

            segment = segments [segId]
            for edgeId in segment.edgeIds:
                e = segments.edgeTable [edgeId]
                if self.Push (e.segmentId, segId, edgeId, dist + e.distance): # internal error - overflow - tables too small
                    expanded += 1

# ================================================================================
